ipa-core/src/protocol/context/dzkp_malicious.rs

```diff
@@ -0,0 +1,151 @@
+use std::{
+    fmt::{Debug, Formatter},
+    num::NonZeroUsize,
+};
+
+use async_trait::async_trait;
+
+use crate::{
+    error::Error,
+    helpers::{ChannelId, Gateway, Message, ReceivingEnd, Role, SendingEnd, TotalRecords},
+    protocol::{
+        context::{
+            dzkp_validator::DZKPBatch, prss::InstrumentedIndexedSharedRandomness, Base,
+            Context as ContextTrait, DZKPContext, InstrumentedSequentialSharedRandomness,
+        },
+        prss::Endpoint as PrssEndpoint,
+        step::{Gate, Step, StepNarrow},
+    },
+    seq_join::SeqJoin,
+    sync::Arc,
+};
+
+/// Represents protocol context in malicious setting, i.e. secure against one active adversary
+/// in 3 party MPC ring.
+#[derive(Clone)]
+pub struct DZKPUpgraded<'a> {
+    /// TODO (alex): Arc is required here because of the `TestWorld` structure. Real world
+    /// may operate with raw references and be more efficient
+    inner: Arc<DZKPUpgradedInner<'a>>,
+    gate: Gate,
+    total_records: TotalRecords,
+}
+
+impl<'a> DZKPUpgraded<'a> {
+    pub(super) fn new<S: Step + ?Sized>(
+        source: &Base<'a>,
+        malicious_step: &S,
```

Comment:
What is this one `Step`? Is this just a way to open a channel for communicating the ZKPs?

Reply:
There is a separate step/gate in DZKPValidator that you can use for the DZKPs and their verification. In the MAC based approach, this was used for the validation as well.